extensions/lifecycle/CircuitBreakerGroup.js

```diff
@@ -0,0 +1,312 @@
+'use-strict';
+/* eslint no-template-curly-in-string: 0 */
+
+const { CircuitBreaker, BreakerState } = require('breakbeat').CircuitBreaker;
+const { startCircuitBreakerMetricsExport } = require('../../lib/CircuitBreaker');
+
+/**
+ * circuit breaker configuration adapted for the bucket processor
+ * @typedef {Object} BucketProcessorCircuitBreakerConfig
+ * @property {Object} global - global circuit breaker configurations (circuit breaks everything)
+ * @property {CircuitBreakerGroup} circuitBreakerGroup - circuit breakers per workflow and location/topic
+ */
+
+class CircuitBreakerGroup {
+    constructor(cbGlobalConf, circuitBreakers) {
+        if (circuitBreakers) {
+            this.circuitBreakers = circuitBreakers;
+            return;
+        }
+
+        this._cbGlobalConf = cbGlobalConf;
+
+        // breaker id used for the metrics
+        this._latestBreakerId = 0;
+
+        this.circuitBreakers = {
+            transition: {
+                topic: {},
+                location: {},
+                global: [],
+            },
+            expiration: {
+                topic: {},
+                location: {},
+                global: [],
+            },
+        };
+    }
+
+    _append(array, element) {
+        if (!array) {
+            return [element];
+        }
+        return [...array, element];
+    }
+
+    _addTransitionCircuitBreaker(circuitBreaker, topic, location) {
+        if (!topic && !location) {
+            this.circuitBreakers.transition.global.push(circuitBreaker);
+        }
+        if (topic) {
+            this.circuitBreakers.transition.topic[topic] =
+                this._append(this.circuitBreakers.transition.topic[topic], circuitBreaker);
+            }
+        if (location) {
+            this.circuitBreakers.transition.location[location] =
+                this._append(this.circuitBreakers.transition.location[location], circuitBreaker);
+        }
+    }
+
+    _addExpirationCircuitBreaker(circuitBreaker, topic, location) {
+        if (!topic && !location) {
+            this.circuitBreakers.expiration.global.push(circuitBreaker);
+        }
+        if (topic) {
+            this.circuitBreakers.expiration.topic[topic] =
+                this._append(this.circuitBreakers.expiration.topic[topic], circuitBreaker);
+        }
+        if (location) {
+            this.circuitBreakers.expiration.location[location] =
+                this._append(this.circuitBreakers.expiration.location[location], circuitBreaker);
+        }
+    }
+
+    exportCircuitBreakerMetric(circuitBreaker, metricsSuffix) {
+        let metricsName = 'lifecycle_bucket_processor';
+        if (metricsSuffix) {
+            metricsName += `_${metricsSuffix}`;
+        }
+        metricsName += `_${this._latestBreakerId}`;
```

Comment:
the 'number' should come first, may be easier to match with the index in the config?